Sample Django Weather Application (#177)
airbrake/pybrake: examples/django/README.md

# Django Sample Application for Pybrake

## About the application:

The example application provides three GET endpoints:

1. `/date` - returns server date and time. 
2. `/locations` - returns list of available locations. 
3. `/weather/{locationName}` - returns the weather details for the locations.

## Steps to run the API:

1. Install the dependencies for the application

```bash

> pip3 install -r requirements.txt

Reviewer: You need to add pybrake to `requirements.txt`. When I do `pip3 install` and then `python manage.py runserver 3000`, I get:

```
ModuleNotFoundError: No module named 'pybrake'
```

```

2. You must set both project_id & project_key in settings.py.

To find your project_id and project_key navigate to your project's Settings and copy the values from the right sidebar.

3. Run the localhost server

```bash
python manage.py runserver 3000
```

4. To retrieve the responses, append the endpoints to the localhost URL with a `/`.

Use the below curl commands to interact with the endpoints. The endpoints require an api-key HTTP header.

```bash

> curl "http://localhost:3000/date/" -H 'api-key: b761be830f7c23ebe1c3250d42c43673' 

Reviewer: ```suggestion
curl "http://localhost:3000/date/"
```

> curl "http://localhost:3000/locations/" -H 'api-key: b761be830f7c23ebe1c3250d42c43673' 

Reviewer: ```suggestion
curl "http://localhost:3000/locations/"
```

> curl "http://localhost:3000/weather/<austin/pune/santabarbara>" -H 'api-key: b761be830f7c23ebe1c3250d42c43673' 

Reviewer: ```suggestion
curl "http://localhost:3000/weather/<austin/pune/santabarbara>"
```

> curl "http://localhost:3000/weather/" -H 'api-key: b761be830f7c23ebe1c3250d42c43673'

Reviewer: ```suggestion
# Should produce an HTTP 404 error
curl -I "http://localhost:3000/weather/"

# Should produce an HTTP 404 error and report the error to Airbrake (since `washington` is not in the supported cities list)
curl -I "http://localhost:3000/weather/washington"
```

```
  

> The last curl command will raise `404 Not Found` error.

Reviewer: ```suggestion

```

# Django Sample Application for Pybrake

## About the application:

The example application provides three GET endpoints:

1. `/date` - returns server date and time. 
2. `/locations` - returns list of available locations. 
3. `/weather/{locationName}` - returns the weather details for the locations.

## Steps to run the API:

1. Install the dependencies for the application

```bash
pip3 install -r requirements.txt
```

2. You must set both project_id & project_key in settings.py.

To find your project_id and project_key navigate to your project's Settings and copy the values from the right sidebar.

3. Run the localhost server

```bash
python manage.py runserver 3000
```

4. To retrieve the responses, append the endpoints to the localhost URL with a `/`.

Use the below curl commands to interact with the endpoints. The endpoints require an api-key HTTP header.

```bash
curl "http://localhost:3000/date/" 
curl "http://localhost:3000/locations/" 
curl "http://localhost:3000/weather/<austin/pune/santabarbara>"

# Should produce an HTTP 404 error
curl "http://localhost:3000/weather/"

> # Should produce an HTTP 404 error and report the error to Airbrake (since `washington` is not in the supported cities list)

Reviewer: ```suggestion
# Should produce an intentional HTTP 500 error and report the error to Airbrake (since `washington` is not in the supported cities list, an `if` condition is bypassed and the `data` variable is used but not initialized)
```

Author: Hey Scott, the response was being returned outside of the try block, fixed that, now we should be getting proper ValueError instead of UnboundLocalError.

# Django Sample Application for Pybrake

## About the application:

The example application provides three GET endpoints:

1. `/date` - returns server date and time. 
2. `/locations` - returns list of available locations. 
3. `/weather/{locationName}` - returns the weather details for the locations.

## Steps to run the API:

1. Install the dependencies for the application

```bash
pip3 install -r requirements.txt
```

> 2. You must set both project_id & project_key in `settings.py`.

Reviewer 2: I am a Django noob, so I had to search for this `settings.py`. Let's change that to `weather/settings.py`.

To find your project_id and project_key from Airbrake account and replace it in below code in your project's `settings.py` file.

```python
AIRBRAKE = dict(
    project_id=999999,                               # Insert your Project Id here
    project_key='xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx',  # Insert your Project Key here
    environment="test",

>     error_notifications=True,                        # False to disable error notification

Reviewer 2: I don't think we want to advertise these options. I would remove lines 28-31. This should work out of the box already.

Reviewer 3: Even `query_stats` and `queue_stats` cannot be disabled by the user?

Added those option over here based on https://github.com/airbrake/pybrake/blob/27c8906e8fe7e6e36faac3c267dbbf0f8184a15e/src/pybrake/notifier.py#L49

https://github.com/airbrake/pybrake/blob/27c8906e8fe7e6e36faac3c267dbbf0f8184a15e/src/pybrake/notifier.py#L50

    performance_stats=True,                          # False to disable APM
    query_stats=True,                                # False to disable query monitoring
    queue_stats=True                                 # False to disable queue monitoring
)
```

3. Run the localhost server

```bash

> python manage.py runserver 3000

Reviewer 3: Updated changed with backward compatible in PR #190.

> python manage.py runserver 3000

Reviewer 2: Approved #190

> python manage.py runserver 3000

Reviewer 3: You should now be able to run it.

```

4. To retrieve the responses, append the endpoints to the localhost URL 
   with a `/`. Use the below curl commands to interact with the endpoints.

```bash
# Working Routes
curl "http://localhost:3000/date" 
curl "http://localhost:3000/locations" 
curl "http://localhost:3000/weather/<austin/pune/santabarbara>"

# Should produce an HTTP 404 error
curl -I "http://localhost:3000/weather"

# Should produce an intentional HTTP 500 error and report the error to Airbrake (since `washington` is not in the supported cities list, an `if` condition is bypassed and the `data` variable is used but not initialized)

> curl -I "http://localhost:3000/weather/washington"

Reviewer 2: `washington` doesn't produce an error 500 for me. I think this has to be looked into.

```
% curl -I "http://localhost:3001/weather/washington"
HTTP/1.1 404 Not Found
Date: Thu, 05 May 2022 09:38:17 GMT
Server: WSGIServer/0.2 CPython/3.9.12
Content-Type: application/json
X-Frame-Options: DENY
Content-Length: 32
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
```

Reviewer 3: Updated, Please check once.
